site/app/repositories/forum/ThreadRepository.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+declare(strict_types=1);
+
+namespace app\repositories\forum;
+
+use app\entities\forum\Thread;
+use Doctrine\ORM\EntityRepository;
+
+class ThreadRepository extends EntityRepository {
+    /**
+     * @param int[] $category_ids
+     * @param int[] $status
+     * @return Thread[]
+     */
+    public function getAllThreads(array $category_ids, array $status, bool $get_deleted, bool $get_merged_threads, bool $filter_unread, string $user_id, int $block_number): array {
+        $block_size = 30;
+        $qb = $this->_em->createQueryBuilder();
+        $qb->select('t')
+            ->from(Thread::class, 't')
+        // AddSelect allows us to join server-side, reducing the number of database queries.
+            ->addSelect('c')
+            ->addSelect('v')
+            ->addSelect('p')
+            ->addSelect('ph')
+            ->addSelect('sf')
+            ->addSelect('u')
+            ->join('t.categories', 'c')
+            ->leftJoin('t.viewers', 'v')
+            ->leftJoin('t.posts', 'p')
+            ->leftJoin('p.history', 'ph')
+            ->leftJoin('t.favorers', 'sf')
+            ->join('t.author', 'u');
+        if (count($category_ids) > 0) {
+            $qb->andWhere('c.category_id IN (:category_ids)')
+                ->setParameter('category_ids', $category_ids);
+        }
+        if (count($status) > 0) {
+            $qb->andWhere('t.status IN (:status)')
+                ->setParameter('status', $status);
+        }
+        if (!$get_deleted) {
+            $qb->andWhere('t.deleted = false');
+        }
+        if (!$get_merged_threads) {
+            $qb->andWhere('t.merged_thread = -1');
+        }
+        $qb->addOrderBy('CASE WHEN t.pinned_expiration > CURRENT_TIMESTAMP() THEN 1 ELSE 0 END', 'DESC')
+            ->addOrderBy('CASE WHEN sf.user_id IS NULL THEN 1 ELSE 0 END', 'DESC')
+            ->addOrderBy('t.id', 'DESC')
+            ->setMaxResults($block_size)
+            ->setFirstResult($block_size * $block_number);
+
+        $result = $qb->getQuery()->getResult();
+
+        if ($filter_unread) {
+            $result = array_filter($result, function ($x) use ($user_id) {
+                return $x->isUnread($user_id);
+            });
+        }
+        return $result;
+    }
+
+
+    public function getThreadDetail(int $thread_id, string $order_posts_by = 'tree'): Thread {
```

Comment:
Would it make sense for `$order_posts_by` to be an enum so it's clear what the valid values are, and so PHPStan can validate that the valid values are covered by any code which uses them?

Reply:
I'm not very familiar with the specifics of PHP enums. This comes straight from the web request's parameter which is text, so if that can be converted to enum easily I can add it. I currently have this as string because that's what the existing forum code does.

Reply:
I guess it's a question of where to put the validation logic.  An enum is more specific, but would also require places where this function is used to convert strings to enum values and deal with the error checking right then and there.